libindy/src/commands/crypto.rs

```diff
@@ -397,15 +455,349 @@ impl CryptoCommandExecutor {
+            })?;
+            let cts_msg = serde_json::to_vec(&cts).map_err(|err| {
+                err_msg(IndyErrorKind::InvalidStructure, format!(
+                    "Failed to serialize msg {}",
+                    err
+                ))
+            })?;
+            Ok((new_msg, cts_msg))
+            //Ok((new_msg, Some(cts_msg)))
+        } else {
+            return Err(err_msg(IndyErrorKind::InvalidStructure, format!(
+                "no {} found", CIPHERTEXT_ARRAY_JSON_KEY
+            )));
+            //Ok((json_msg, None))
+        }
+    }
+
+    /// Accepts a message and content of `~cyphertexts` and returns a message with `~cyphertexts` key
+    /// present. If `~cyphertexts` is already present, an error is returned.
+    pub fn add_cts_to_msg(json_msg: &[u8], json_cts: &[u8]) -> IndyResult<Vec<u8>> {
+        let mut json: Value = serde_json::from_slice(&json_msg).map_err(|err| {
+            err_msg(IndyErrorKind::InvalidStructure, format!(
+                "Failed to deserialize message to json {}",
+                err
+            ))
+        })?;
+
+        let msg_obj = match json.as_object_mut() {
+            Some(o) => o,
+            None => return Err(err_msg(IndyErrorKind::InvalidStructure, format!(
+                "Expected an object"
+            )))
+        };
+
+        if msg_obj.contains_key(CIPHERTEXT_ARRAY_JSON_KEY) {
+            return Err(err_msg(IndyErrorKind::InvalidStructure, format!(
+                "Already contains {}", CIPHERTEXT_ARRAY_JSON_KEY
+            )));
+        } else {
+            let json_cts: Value = serde_json::from_slice(&json_cts).map_err(|err| {
+                err_msg(IndyErrorKind::InvalidStructure, format!(
+                    "Failed to deserialize message to json {}",
+                    err
+                ))
+            })?;
+            msg_obj.insert(CIPHERTEXT_ARRAY_JSON_KEY.to_string(), json_cts);
+        }
+
+        serde_json::to_vec(&msg_obj).map_err(|err| {
+            err_msg(IndyErrorKind::InvalidStructure, format!(
+                "Failed to serialize {}",
+                err
+            ))
+        })
+    }
+    /// For pre-processing of a message before unpack. Takes a message which is the result of
+    /// `collapse_ciphertext` and returns a message where the placeholders have been replaced with
+    /// original values which are elements of the last object of `~cyphertexts` and that last object
+    /// is removed from `~cyphertexts`.
+    pub fn expand_ciphertext(packed_msg: &[u8]) -> IndyResult<Vec<u8>> {
+        let jwe_cd_struct: JWEWithCD = serde_json::from_slice(&packed_msg).map_err(|err| {
+            err_msg(IndyErrorKind::InvalidStructure, format!(
+                "Failed to deserialize message to JWEWithCD {}",
+                err
+            ))
+        })?;
+
+        let (iv, tag, ciphertext) = (jwe_cd_struct.iv, jwe_cd_struct.tag, jwe_cd_struct.ciphertext);
+
+        let b1 = IDX_REGEX.is_match(&iv);
```

Comment:
Please improve readability of this block of code. Most probably it's possible to use serde instead of ugly match (b1, b2, b3)